sdks/python/apache_beam/transforms/periodicsequence.py

```diff
@@ -173,31 +216,92 @@ class PeriodicImpulse(PTransform):
   but can be used as first transform in pipeline.
   The PCollection generated by PeriodicImpulse is unbounded.
   '''
+  def _validate_and_adjust_duration(self):
+    assert self.data
+
+    # The total time we need to impulse all the data.
+    data_duration = (len(self.data) - 1) * self.interval
+
+    is_pre_timestamped = isinstance(self.data[0], tuple) and \
+      isinstance(self.data[0][0], timestamp.Timestamp)
+
+    if isinstance(self.start_ts, Timestamp):
+      start = self.start_ts.micros / 1000000
+    else:
+      start = self.start_ts
+
+    if isinstance(self.stop_ts, Timestamp):
+      if self.stop_ts == MAX_TIMESTAMP:
+        # adjust stop timestamp to match the data duration
+        end = start + data_duration
+        if self.interval > 1e-6:
+          end += 1e-6
+        self.stop_ts = Timestamp.of(end)
+      else:
+        end = self.stop_ts.micros / 1000000
+    else:
+      end = self.stop_ts
+
+    # The total time for the impulse signal which occurs in [start, end).
+    impulse_duration = end - start
+    if data_duration + self.interval < impulse_duration:
+      # We don't have enough data for the impulse.
+      # If we can fit at least one more data point in the impulse duration,
+      # then we will be in the repeat mode.
+      message = 'The number of elements in the provided pre-timestamped ' \
+        'data sequence is not enough to span the full impulse duration. ' \
+        f'Expected duration: {impulse_duration:.6f}, ' \
+        f'actual data duration: {data_duration:.6f}.'
+
+      if is_pre_timestamped:
+        raise ValueError(
+            f'{message} Please either provide more data or decrease '
+            '`stop_timestamp`.')
+      else:
+        warnings.warn(
+            f'{message} As a result, the data sequence will be repeated to '
+            'generate elements for the entire duration.')
+
   def __init__(
       self,
-      start_timestamp=Timestamp.now(),
-      stop_timestamp=MAX_TIMESTAMP,
-      fire_interval=360.0,
-      apply_windowing=False):
+      start_timestamp: Union[Timestamp, float] = Timestamp.now(),
+      stop_timestamp: Union[Timestamp, float] = MAX_TIMESTAMP,
```

Comment:
I'm fine with allowing floats here, but could you explain what float means in the pydoc? Alternatively, we can just keep Timestamp as the type.